File: core/core-api/src/main/java/com/coffee/gu/support/event/InternalEventRecoveryScheduler.java
package com.coffee.gu.support.event;

import com.coffee.gu.EventDispatcher;
import com.coffee.gu.EventRecoveryScheduler;
import com.coffee.gu.enums.EventLogTarget;
import com.coffee.gu.EventLog;
import com.coffee.gu.EventLogRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class InternalEventRecoveryScheduler implements EventRecoveryScheduler {

    private static final Logger log = LoggerFactory.getLogger(InternalEventRecoveryScheduler.class);
    private static final int MAX_RETRY_LIMIT = 5;

    private final EventLogRepository eventLogRepository;
    private final List<EventDispatcher> eventDispatchers;

    public InternalEventRecoveryScheduler(EventLogRepository eventLogRepository, List<EventDispatcher> eventDispatchers) {
        this.eventLogRepository = eventLogRepository;
        this.eventDispatchers = eventDispatchers;
    }

    @Scheduled(fixedRate = 5000)
    public void republish() {
        List<EventLog> eventLogs = eventLogRepository.getUnpublishedEventLogs(EventLogTarget.INTERNAL);
        for (EventLog eventLog : eventLogs) {
            if (eventLog.getRetryCount() != null && eventLog.getRetryCount() >= MAX_RETRY_LIMIT) {
                log.warn("Outbox 이벤트 재발행 5회 초과 실패로 DEAD 전환: eventId={}, eventType={}", eventLog.getEventId(), eventLog.getEventType());
                eventLogRepository.markAsDead(eventLog.getEventId());
                // 알림호출
                continue;
            }

            try {
                eventDispatchers.stream()
                        .filter(eventDispatcher -> eventDispatcher.supports(eventLog.getEventType()))
                        .forEach(eventDispatcher -> eventDispatcher.dispatch(eventLog.getPayload()));

                eventLogRepository.republish(List.of(eventLog.getEventId()));
            } catch (Exception e) {
                log.error("Outbox 이벤트 재발행 실패 (retryCount 증가): eventId={}", eventLog.getEventId(), e);
                eventLogRepository.increaseRetryCount(eventLog.getEventId());
            }
        }
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/support/event/InternalEventRecoveryScheduler.kt
package com.coffee.gu.support.event

import com.coffee.gu.EventDispatcher
import com.coffee.gu.EventLogRepository
import com.coffee.gu.EventRecoveryScheduler
import com.coffee.gu.enums.EventLogTarget
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

@Component
class InternalEventRecoveryScheduler(
    private val eventLogRepository: EventLogRepository,
    private val eventDispatchers: List<EventDispatcher>,
) : EventRecoveryScheduler {

    companion object {
        private val log = LoggerFactory.getLogger(InternalEventRecoveryScheduler::class.java)
        private const val MAX_RETRY_LIMIT = 5
    }

    @Scheduled(fixedRate = 5000)
    override fun republish() {
        val eventLogs = eventLogRepository.getUnpublishedEventLogs(EventLogTarget.INTERNAL)
        for (eventLog in eventLogs) {
            if (eventLog.retryCount != null && eventLog.retryCount!! >= MAX_RETRY_LIMIT) {
                log.warn("Outbox 이벤트 재발행 5회 초과 실패로 DEAD 전환: eventId={}, eventType={}", eventLog.eventId, eventLog.eventType)
                eventLogRepository.markAsDead(eventLog.eventId)
                continue
            }

            try {
                eventDispatchers.filter { it.supports(eventLog.eventType) }
                    .forEach { it.dispatch(eventLog.payload) }

                eventLogRepository.republish(listOf(eventLog.eventId))
            } catch (e: Exception) {
                log.error("Outbox 이벤트 재발행 실패 (retryCount 증가): eventId={}", eventLog.eventId, e)
                eventLogRepository.increaseRetryCount(eventLog.eventId)
            }
        }
    }
}
